File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/general/StringAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import java.util.List;

import com.truveta.opentoken.attributes.BaseAttribute;

/**
 * Represents a generic string attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * general text fields. It recognizes "String" and "Text" as valid aliases for
 * this attribute type.
 * 
 * The attribute performs normalization on input values by trimming leading and
 * trailing whitespace.
 * 
 * The attribute validates that values are not null or empty (after trimming).
 */
public class StringAttribute extends BaseAttribute {

    private static final String NAME = "String";
    private static final String[] ALIASES = new String[] { NAME, "Text" };

    public StringAttribute() {
        super(List.of());
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        if (value == null) {
            throw new IllegalArgumentException("String value cannot be null");
        }

        return value.trim();
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/general/string_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List

from opentoken.attributes.base_attribute import BaseAttribute


class StringAttribute(BaseAttribute):
    """Represents a generic string attribute.

    This class extends BaseAttribute and provides functionality for working with
    string fields. It normalizes values by trimming whitespace.

    The attribute validates that the value is not null or empty using the
    default BaseAttribute validation rules.
    """

    NAME = "String"
    ALIASES = [NAME]

    def __init__(self):
        # Use default validation rules from BaseAttribute (not null or empty)
        super().__init__()

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "String"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize the string value by trimming whitespace.

        Args:
            value: The string value to normalize

        Returns:
            str: The trimmed string value
        """
        return value.strip()
